File: session/wechat_bot/media.py
"""Media download and upload for WeChat.

Handles:
  - Downloading + AES decrypting inbound media (images, voice, files, videos)
  - AES encrypting + uploading outbound media to WeChat CDN

Protocol (from @tencent-weixin/openclaw-weixin v1.0.3):
  Download: GET CDN url -> AES-128-ECB decrypt -> save to disk
  Upload:   AES-128-ECB encrypt -> POST to CDN -> get download param -> sendmessage
"""

from __future__ import annotations

import base64
import hashlib
import os
import time
from pathlib import Path
from typing import Any
from urllib.parse import quote

import httpx

from . import crypto

# File extension sets
_IMAGE_EXTS = {".jpg", ".jpeg", ".png", ".gif", ".bmp", ".webp", ".tiff", ".ico", ".svg"}
_VIDEO_EXTS = {".mp4", ".avi", ".mov", ".mkv", ".webm", ".flv"}
_VOICE_EXTS = {".mp3", ".wav", ".amr", ".silk", ".ogg", ".m4a", ".aac", ".flac"}

# Protocol constants
_UPLOAD_MEDIA_IMAGE = 1
_UPLOAD_MEDIA_VIDEO = 2
_UPLOAD_MEDIA_FILE = 3
_UPLOAD_MEDIA_VOICE = 4

_ITEM_TEXT = 1
_ITEM_IMAGE = 2
_ITEM_VOICE = 3
_ITEM_FILE = 4
_ITEM_VIDEO = 5

_MESSAGE_TYPE_BOT = 2
_MESSAGE_STATE_FINISH = 2


def _has_download_url(media: dict[str, Any] | None) -> bool:
    """Check if a media dict has downloadable URL data."""
    if not isinstance(media, dict):
        return False
    eqp = str(media.get("encrypt_query_param", "") or "").strip()
    full = str(media.get("full_url", "") or "").strip()
    return bool(eqp or full)


def _ext_for_type(media_type: str) -> str:
    """Get file extension for a media type."""
    return {"image": ".jpg", "voice": ".silk", "video": ".mp4", "file": ""}.get(
        media_type, ""
    )


def _guess_upload_type(filename: str) -> tuple[int, int, str]:
    """Determine upload media type, item type, and item key from filename extension.

    Returns:
        (upload_media_type, item_type, item_key)
    """
    ext = Path(filename).suffix.lower()
    if ext in _IMAGE_EXTS:
        return _UPLOAD_MEDIA_IMAGE, _ITEM_IMAGE, "image_item"
    if ext in _VIDEO_EXTS:
        return _UPLOAD_MEDIA_VIDEO, _ITEM_VIDEO, "video_item"
    if ext in _VOICE_EXTS:
        return _UPLOAD_MEDIA_VOICE, _ITEM_VOICE, "voice_item"
    return _UPLOAD_MEDIA_FILE, _ITEM_FILE, "file_item"


async def download_media(
    client: httpx.AsyncClient,
    cdn_base_url: str,
    typed_item: dict[str, Any],
    media_type: str,
    save_dir: Path,
    filename: str | None = None,
) -> str | None:
    """Download and decrypt a media item from WeChat CDN.

    Args:
        client: HTTP client for requests.
        cdn_base_url: CDN base URL.
        typed_item: The typed item dict (e.g. image_item, file_item).
        media_type: One of "image", "voice", "file", "video".
        save_dir: Directory to save downloaded files.
        filename: Optional filename override.

    Returns:
        Local file path on success, None on failure.
    """
    try:
        media = typed_item.get("media") or {}
        encrypt_query_param = str(media.get("encrypt_query_param", "") or "").strip()
        full_url = str(media.get("full_url", "") or "").strip()

        if not encrypt_query_param and not full_url:
            return None

        # Resolve AES key
        # image_item.aeskey is raw hex (32 chars) -> convert to base64
        # media.aes_key is always base64
        raw_aeskey_hex = typed_item.get("aeskey", "")
        media_aes_key_b64 = media.get("aes_key", "")

        aes_key_b64 = ""
        if raw_aeskey_hex:
            aes_key_b64 = base64.b64encode(bytes.fromhex(raw_aeskey_hex)).decode()
        elif media_aes_key_b64:
            aes_key_b64 = media_aes_key_b64

        # Non-image media requires AES key
        if media_type != "image" and not aes_key_b64:
            return None

        # Build download URL candidates
        fallback_url = ""
        if encrypt_query_param:
            fallback_url = (
                f"{cdn_base_url}/download"
                f"?encrypted_query_param={quote(encrypt_query_param)}"
            )

        candidates: list[tuple[str, str]] = []
        if full_url:
            candidates.append(("full_url", full_url))
        if fallback_url and (not full_url or fallback_url != full_url):
            candidates.append(("encrypt_query_param", fallback_url))

        # Try each URL
        data = b""
        for source, cdn_url in candidates:
            try:
                resp = await client.get(cdn_url)
                resp.raise_for_status()
                data = resp.content
                break
            except Exception:
                continue

        if not data:
            return None

        # Decrypt if key available
        if aes_key_b64:
            data = crypto.decrypt(data, aes_key_b64)

        # Save to disk
        save_dir.mkdir(parents=True, exist_ok=True)
        if not filename:
            ts = int(time.time())
            h = abs(hash(encrypt_query_param or full_url)) % 100000
            filename = f"{media_type}_{ts}_{h}{_ext_for_type(media_type)}"
        safe_name = os.path.basename(filename)
        file_path = save_dir / safe_name
        file_path.write_bytes(data)
        return str(file_path)

    except Exception:
        return None


async def upload_and_send_media(
    client: httpx.AsyncClient,
    api_post,  # callable: async (endpoint, body) -> dict
    base_url: str,
    cdn_base_url: str,
    to_user_id: str,
    media_path: str,
    context_token: str,
) -> bool:
    """Upload a local file to WeChat CDN and send it as a media message.

    Protocol:
      1. Generate random 16-byte AES key
      2. Call getuploadurl with file metadata + AES key
      3. AES-128-ECB encrypt file data and POST to CDN
      4. Read x-encrypted-param from CDN response
      5. Send sendmessage with media item referencing the upload

    Args:
        client: HTTP client.
        api_post: Async callable for API POST requests.
        base_url: API base URL.
        cdn_base_url: CDN base URL.
        to_user_id: Recipient user ID.
        media_path: Local file path to upload.
        context_token: Context token for the conversation.

    Returns:
        True on success, False on failure.
    """
    try:
        p = Path(media_path)
        if not p.is_file():
            return False

        raw_data = p.read_bytes()
        raw_size = len(raw_data)
        raw_md5 = hashlib.md5(raw_data).hexdigest()

        upload_type, item_type, item_key = _guess_upload_type(p.name)

        # Client-side AES key
        aes_key_raw = os.urandom(16)
        aes_key_hex = aes_key_raw.hex()

        # PKCS7 padded size
        padded_size = ((raw_size + 1 + 15) // 16) * 16

        # Step 1: Get upload URL
        file_key = os.urandom(16).hex()
        upload_resp = await api_post("ilink/bot/getuploadurl", {
            "filekey": file_key,
            "media_type": upload_type,
            "to_user_id": to_user_id,
            "rawsize": raw_size,
            "rawfilemd5": raw_md5,
            "filesize": padded_size,
            "no_need_thumb": True,
            "aeskey": aes_key_hex,
        })

        upload_full_url = str(upload_resp.get("upload_full_url", "") or "").strip()
        upload_param = str(upload_resp.get("upload_param", "") or "")
        if not upload_full_url and not upload_param:
            return False

        # Step 2: Encrypt and upload to CDN
        aes_key_b64 = base64.b64encode(aes_key_raw).decode()
        encrypted_data = crypto.encrypt(raw_data, aes_key_b64)

        if upload_full_url:
            cdn_url = upload_full_url
        else:
            cdn_url = (
                f"{cdn_base_url}/upload"
                f"?encrypted_query_param={quote(upload_param)}"
                f"&filekey={quote(file_key)}"
            )

        cdn_resp = await client.post(
            cdn_url,
            content=encrypted_data,
            headers={"Content-Type": "application/octet-stream"},
        )
        cdn_resp.raise_for_status()

        download_param = cdn_resp.headers.get("x-encrypted-param", "")
        if not download_param:
            return False

        # Step 3: Send message with media reference
        cdn_aes_key_b64 = base64.b64encode(aes_key_hex.encode()).decode()

        import uuid

        media_item: dict[str, Any] = {
            "media": {
                "encrypt_query_param": download_param,
                "aes_key": cdn_aes_key_b64,
                "encrypt_type": 1,
            },
        }

        if item_type == _ITEM_IMAGE:
            media_item["mid_size"] = padded_size
        elif item_type == _ITEM_VIDEO:
            media_item["video_size"] = padded_size
        elif item_type == _ITEM_FILE:
            media_item["file_name"] = p.name
            media_item["len"] = str(raw_size)

        item_list = [{"type": item_type, item_key: media_item}]
        weixin_msg: dict[str, Any] = {
            "from_user_id": "",
            "to_user_id": to_user_id,
            "client_id": f"bot-{uuid.uuid4().hex[:12]}",
            "message_type": _MESSAGE_TYPE_BOT,
            "message_state": _MESSAGE_STATE_FINISH,
            "item_list": item_list,
        }
        if context_token:
            weixin_msg["context_token"] = context_token

        await api_post("ilink/bot/sendmessage", {"msg": weixin_msg})
        return True

    except Exception:
        return False

File: webot/src-tauri/src/media.rs
use aes::cipher::{BlockEncrypt, KeyInit};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine};
use serde_json::Value;

pub const CDN_BASE_URL: &str = "https://novac2c.cdn.weixin.qq.com/c2c";

const UPLOAD_MEDIA_IMAGE: u32 = 1;
const UPLOAD_MEDIA_VIDEO: u32 = 2;
const UPLOAD_MEDIA_FILE: u32 = 3;
const UPLOAD_MEDIA_VOICE: u32 = 4;

const ITEM_IMAGE: u32 = 2;
const ITEM_VOICE: u32 = 3;
const ITEM_FILE: u32 = 4;
const ITEM_VIDEO: u32 = 5;

const IMAGE_EXTS: &[&str] = &[
    ".jpg", ".jpeg", ".png", ".gif", ".bmp", ".webp", ".tiff", ".ico", ".svg",
];
const VIDEO_EXTS: &[&str] = &[".mp4", ".avi", ".mov", ".mkv", ".webm", ".flv"];
const VOICE_EXTS: &[&str] = &[
    ".mp3", ".wav", ".amr", ".silk", ".ogg", ".m4a", ".aac", ".flac",
];

struct MediaTypeInfo {
    upload_type: u32,
    item_type: u32,
    item_key: &'static str,
}

fn detect_media_type(path: &std::path::Path) -> MediaTypeInfo {
    let ext = path.extension().and_then(|e| e.to_str()).unwrap_or("");
    let ext_lower = format!(".{}", ext.to_lowercase());

    if IMAGE_EXTS.contains(&ext_lower.as_str()) {
        MediaTypeInfo {
            upload_type: UPLOAD_MEDIA_IMAGE,
            item_type: ITEM_IMAGE,
            item_key: "image_item",
        }
    } else if VIDEO_EXTS.contains(&ext_lower.as_str()) {
        MediaTypeInfo {
            upload_type: UPLOAD_MEDIA_VIDEO,
            item_type: ITEM_VIDEO,
            item_key: "video_item",
        }
    } else if VOICE_EXTS.contains(&ext_lower.as_str()) {
        MediaTypeInfo {
            upload_type: UPLOAD_MEDIA_VOICE,
            item_type: ITEM_VOICE,
            item_key: "voice_item",
        }
    } else {
        MediaTypeInfo {
            upload_type: UPLOAD_MEDIA_FILE,
            item_type: ITEM_FILE,
            item_key: "file_item",
        }
    }
}

fn md5_hex(data: &[u8]) -> String {
    use md5::{Md5, Digest};
    let mut hasher = Md5::new();
    hasher.update(data);
    hex::encode(hasher.finalize())
}

fn aes_ecb_encrypt(data: &[u8], key: &[u8; 16]) -> Vec<u8> {
    let pad_len = 16 - (data.len() % 16);
    let mut buf = data.to_vec();
    buf.extend(std::iter::repeat(pad_len as u8).take(pad_len));

    let cipher = aes::Aes128::new(key.into());
    for chunk in buf.chunks_exact_mut(16) {
        let mut block = *aes::cipher::generic_array::GenericArray::from_slice(
            chunk as &[u8],
        );
        cipher.encrypt_block(&mut block);
        chunk.copy_from_slice(&block);
    }
    buf
}

async fn get_upload_url(
    client: &reqwest::Client,
    base_url: &str,
    auth_headers: &reqwest::header::HeaderMap,
    file_key: &str,
    upload_type: u32,
    to_user_id: &str,
    raw_size: usize,
    raw_md5: &str,
    padded_size: usize,
    aes_key_hex: &str,
    base_info: &Value,
) -> Result<(String, String), String> {
    let body = serde_json::json!({
        "filekey": file_key,
        "media_type": upload_type,
        "to_user_id": to_user_id,
        "rawsize": raw_size,
        "rawfilemd5": raw_md5,
        "filesize": padded_size,
        "no_need_thumb": true,
        "aeskey": aes_key_hex,
        "base_info": base_info,
    });

    let resp = client
        .post(format!("{}/ilink/bot/getuploadurl", base_url))
        .headers(auth_headers.clone())
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("getuploadurl 请求失败: {e}"))?;

    let data: Value = resp.json().await.map_err(|e| format!("解析 getuploadurl 响应失败: {e}"))?;

    let upload_full_url = data
        .get("upload_full_url")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_string();
    let upload_param = data
        .get("upload_param")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();

    if upload_full_url.is_empty() && upload_param.is_empty() {
        return Err(format!(
            "getuploadurl 未返回上传地址: {}",
            data.get("errmsg").and_then(|v| v.as_str()).unwrap_or("unknown")
        ));
    }

    Ok((upload_full_url, upload_param))
}

async fn upload_to_cdn(
    client: &reqwest::Client,
    upload_full_url: &str,
    upload_param: &str,
    file_key: &str,
    encrypted_data: &[u8],
) -> Result<String, String> {
    let url = if !upload_full_url.is_empty() {
        upload_full_url.to_string()
    } else {
        format!(
            "{}/upload?encrypted_query_param={}&filekey={}",
            CDN_BASE_URL,
            urlencoding(upload_param),
            urlencoding(file_key),
        )
    };

    let resp = client
        .post(&url)
        .header("Content-Type", "application/octet-stream")
        .body(encrypted_data.to_vec())
        .send()
        .await
        .map_err(|e| format!("CDN 上传失败: {e}"))?;

    let download_param = resp
        .headers()
        .get("x-encrypted-param")
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();

    if download_param.is_empty() {
        return Err(format!(
            "CDN 上传响应缺少 x-encrypted-param, status={}",
            resp.status()
        ));
    }

    Ok(download_param)
}

fn urlencoding(s: &str) -> String {
    s.chars()
        .map(|c| match c {
            'A'..='Z' | 'a'..='z' | '0'..='9' | '-' | '_' | '.' | '~' => c.to_string(),
            _ => format!("%{:02X}", c as u8),
        })
        .collect()
}

fn build_media_message_body(
    to_user_id: &str,
    context_token: &str,
    type_info: &MediaTypeInfo,
    download_param: &str,
    aes_key_hex: &str,
    padded_size: usize,
    raw_size: usize,
    file_name: &str,
    base_info: &Value,
) -> Value {
    let aes_key_b64 = BASE64.encode(aes_key_hex.as_bytes());

    let media = serde_json::json!({
        "encrypt_query_param": download_param,
        "aes_key": aes_key_b64,
        "encrypt_type": 1,
    });

    let item_content = match type_info.item_key {
        "image_item" => serde_json::json!({
            "media": media,
            "mid_size": padded_size,
        }),
        "video_item" => serde_json::json!({
            "media": media,
            "video_size": padded_size,
        }),
        "voice_item" => serde_json::json!({
            "media": media,
        }),
        _ => serde_json::json!({
            "media": media,
            "file_name": file_name,
            "len": raw_size.to_string(),
        }),
    };

    let item = serde_json::json!({
        "type": type_info.item_type,
        type_info.item_key: item_content,
    });

    serde_json::json!({
        "msg": {
            "from_user_id": "",
            "to_user_id": to_user_id,
            "client_id": format!("webot-{}", &uuid::Uuid::new_v4().to_string()[..12]),
            "message_type": 2,
            "message_state": 2,
            "item_list": [item],
            "context_token": context_token,
        },
        "base_info": base_info,
    })
}

pub async fn send_media_file(
    client: &reqwest::Client,
    base_url: &str,
    auth_headers: &reqwest::header::HeaderMap,
    to_user_id: &str,
    context_token: &str,
    file_path: &str,
    base_info: &Value,
) -> Result<(), String> {
    let path = std::path::Path::new(file_path);
    if !path.is_file() {
        return Err(format!("文件不存在: {file_path}"));
    }

    let raw_data = std::fs::read(path).map_err(|e| format!("读取文件失败: {e}"))?;
    let raw_size = raw_data.len();
    let raw_md5 = md5_hex(&raw_data);
    let type_info = detect_media_type(path);

    let aes_key_bytes: [u8; 16] = *uuid::Uuid::new_v4().as_bytes();
    let aes_key_hex = hex::encode(aes_key_bytes);
    let file_key = uuid::Uuid::new_v4().simple().to_string();

    let padded_size = ((raw_size + 1 + 15) / 16) * 16;
    let encrypted = aes_ecb_encrypt(&raw_data, &aes_key_bytes);

    let (upload_full_url, upload_param) = get_upload_url(
        client,
        base_url,
        auth_headers,
        &file_key,
        type_info.upload_type,
        to_user_id,
        raw_size,
        &raw_md5,
        padded_size,
        &aes_key_hex,
        base_info,
    )
    .await?;

    let download_param = upload_to_cdn(
        client,
        &upload_full_url,
        &upload_param,
        &file_key,
        &encrypted,
    )
    .await?;

    let msg_body = build_media_message_body(
        to_user_id,
        context_token,
        &type_info,
        &download_param,
        &aes_key_hex,
        padded_size,
        raw_size,
        path.file_name().and_then(|n| n.to_str()).unwrap_or("file"),
        base_info,
    );

    let resp = client
        .post(format!("{}/ilink/bot/sendmessage", base_url))
        .headers(auth_headers.clone())
        .json(&msg_body)
        .send()
        .await
        .map_err(|e| format!("sendmessage 请求失败: {e}"))?;

    let data: Value = resp.json().await.map_err(|e| format!("解析 sendmessage 响应失败: {e}"))?;
    let errcode = data.get("errcode").and_then(|v| v.as_i64()).unwrap_or(0);
    if errcode != 0 {
        return Err(format!(
            "sendmessage 失败: errcode={}, {}",
            errcode,
            data.get("errmsg").and_then(|v| v.as_str()).unwrap_or("")
        ));
    }

    Ok(())
}
